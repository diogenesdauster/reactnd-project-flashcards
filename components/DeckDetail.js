import React, { Component } from "react";
import PropTypes from "prop-types";
import {
  View,
  Text,
  TouchableOpacity,
  StyleSheet,
  Alert,
  Button
} from "react-native";
import { lightblue3, gray, white, black } from "../utils/colors";

class DeckDetail extends Component {
  static propTypes = {
    navigation: PropTypes.shape({
      navigate: PropTypes.func.isRequired,
      state: PropTypes.shape({
        params: PropTypes.shape({
          title: PropTypes.string.isRequired
        }).isRequired
      }).isRequired
    }).isRequired,
    deck: PropTypes.shape({
      title: PropTypes.string.isRequired,
      questions: PropTypes.array.isRequired
    }).isRequired,
    getDeckByTitle: PropTypes.func.isRequired
  };

  static navigationOptions = ({ navigation }) => {
    return {
      headerLeft: (
        <Button
          onPress={() => {
            navigation.popToTop();
          }}
          color={white}
          title="Home"
        />
      )
    };
  };

  state = {
    ready: false
  };

  componentDidMount() {
    const { getDeckByTitle, navigation } = this.props;
    const { title } = navigation.state.params;

    getDeckByTitle(title, () => this.setState({ ready: true }));
  }

  _handleAddNewCard = () => {
    const {
      navigation: { navigate },
      deck: { title }
    } = this.props;
    navigate("AddCard", { title });
  };

  _handleStartQuiz = () => {
    const { navigate } = this.props.navigation;
    const { questions } = this.props.deck;
    if (questions.length) {
      navigate("Quiz", { deck: this.props.deck });
    } else {
      Alert.alert(
        "No card found",
        "Please add card to the deck to enable the quiz"
      );
    }
  };
  render() {
    const { title, questions } = this.props.deck;
    return (
      <View style={styles.container}>
        <Text style={styles.title}>{title}</Text>
        <Text style={styles.cardCount}>
          {questions ? questions.length : 0}{" "}
          {questions && questions.length > 1 ? "cards" : "card"}
        </Text>
        <TouchableOpacity
          onPress={this._handleAddNewCard}
          style={styles.addCardBtn}
        >
          <Text style={{ fontSize: 32, color: white }}>Add Card</Text>
        </TouchableOpacity>
        <TouchableOpacity
          onPress={this._handleStartQuiz}
          style={styles.startQuizBtn}
        >
          <Text style={{ fontSize: 32, color: black }}>Start Quiz</Text>
        </TouchableOpacity>
      </View>
    );
  }
}

export default DeckDetail;

const styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: "center",
    alignItems: "center"
  },
  title: {
    fontSize: 56,
    fontWeight: "bold"
  },
  cardCount: {
    fontSize: 36,
    color: gray,
    marginBottom: 60
  },
  addCardBtn: {
    width: 250,
    height: 60,
    marginBottom: 10,
    backgroundColor: lightblue3,
    justifyContent: "center",
    alignItems: "center",
    shadowRadius: 3,
    shadowOpacity: 0.8,
    shadowColor: "rgba(0, 0, 0, 0.24)",
    shadowOffset: {
      width: 0,
      height: 3
    }
  },
  startQuizBtn: {
    width: 250,
    height: 60,
    backgroundColor: white,
    justifyContent: "center",
    alignItems: "center",
    shadowRadius: 3,
    shadowOpacity: 0.8,
    shadowColor: "rgba(0, 0, 0, 0.24)",
    shadowOffset: {
      width: 0,
      height: 3
    }
  }
});
