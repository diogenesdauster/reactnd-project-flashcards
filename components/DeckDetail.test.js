import "react-native";
import React from "react";
import DeckDetail from "./DeckDetail";
// Note: test renderer must be required after react-native.
import renderer from "react-test-renderer";

// static propTypes = {
//     navigation: PropTypes.shape({
//       navigate: PropTypes.func.isRequired,
//       state: PropTypes.shape({
//         params: PropTypes.shape({
//           title: PropTypes.string.isRequired
//         }).isRequired
//       }).isRequired
//     }).isRequired,
//     deck: PropTypes.shape({
//       title: PropTypes.string.isRequired,
//       questions: PropTypes.array.isRequired
//     }).isRequired,
//     getDeckByTitle: PropTypes.func.isRequired
//   };

it("renders DeckDetail correctly", () => {
  const getDeckByTitle = jest.fn();
  const mockNavigation = {
    navigate: jest.fn(),
    state: { params: { title: "CardTest" } }
  };
  const mockDeck = {
    title: "React",
    questions: [
      {
        question: "What is React?",
        answer: "A library for managing user interfaces"
      },
      {
        question: "Where do you make Ajax requests in React?",
        answer: "The componentDidMount lifecycle event"
      }
    ]
  };

  const tree = renderer
    .create(
      <DeckDetail
        navigation={mockNavigation}
        deck={mockDeck}
        getDeckByTitle={getDeckByTitle}
      />
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
