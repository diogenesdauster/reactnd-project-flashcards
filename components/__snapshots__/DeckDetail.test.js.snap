// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders DeckDetail correctly 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    style={
      Object {
        "fontSize": 56,
        "fontWeight": "bold",
      }
    }
  >
    React
  </Text>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    style={
      Object {
        "color": "#757575",
        "fontSize": 36,
        "marginBottom": 60,
      }
    }
  >
    2
     
    cards
  </Text>
  <View
    accessibilityComponentType={undefined}
    accessibilityLabel={undefined}
    accessibilityTraits={undefined}
    accessible={true}
    collapsable={undefined}
    hasTVPreferredFocus={undefined}
    hitSlop={undefined}
    isTVSelectable={true}
    nativeID={undefined}
    onLayout={undefined}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#00BCD4",
        "height": 60,
        "justifyContent": "center",
        "marginBottom": 10,
        "opacity": 1,
        "shadowColor": "rgba(0, 0, 0, 0.24)",
        "shadowOffset": Object {
          "height": 3,
          "width": 0,
        },
        "shadowOpacity": 0.8,
        "shadowRadius": 3,
        "width": 250,
      }
    }
    testID={undefined}
    tvParallaxProperties={undefined}
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "color": "#FFFFFF",
          "fontSize": 32,
        }
      }
    >
      Add Card
    </Text>
  </View>
  <View
    accessibilityComponentType={undefined}
    accessibilityLabel={undefined}
    accessibilityTraits={undefined}
    accessible={true}
    collapsable={undefined}
    hasTVPreferredFocus={undefined}
    hitSlop={undefined}
    isTVSelectable={true}
    nativeID={undefined}
    onLayout={undefined}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#FFFFFF",
        "height": 60,
        "justifyContent": "center",
        "opacity": 1,
        "shadowColor": "rgba(0, 0, 0, 0.24)",
        "shadowOffset": Object {
          "height": 3,
          "width": 0,
        },
        "shadowOpacity": 0.8,
        "shadowRadius": 3,
        "width": 250,
      }
    }
    testID={undefined}
    tvParallaxProperties={undefined}
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "color": undefined,
          "fontSize": 32,
        }
      }
    >
      Start Quiz
    </Text>
  </View>
</View>
`;
